### ISSUE TITLE
It seems that `recompute_grad` is not saving GPU memory.

### ISSUE THREAD

**Issue by nyngwang** [tensorflow/tensorflow#57205]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

conda `tensorflow-gpu` 2.10.0rc1 pypi_0 pypi
(I also tried: conda `tensorflow` 2.9.1   pypi_0    pypi)

### Custom Code

No

### OS Platform and Distribution

Linux Ubuntu 20.04

### Python version

3.10.5

### CUDA/cuDNN version

cudatoolkit  11.2.2   hbe64b41_10    conda-forge
cudnn          8.1.0.77   h90431f1_0    conda-forge

### GPU model and memory

NVIDIA GeForce RTX 3090 24GiB

### Current Behaviour?

(originally from this thread: https://github.com/davisyoshida/tf2-gradient-checkpointing/issues/1)

I cannot tell whether the APIs (either the official one `tf.recompute_grad`, or those from GitHub by [@pidajay](https://github.com/pidajay/tf2_gradient_checkpointing), [someone from google - @ppham27](https://github.com/tensorflow/tensorflow/issues/30418#issuecomment-719124842), [someone that crop tf-slim - @mathemakitten](https://github.com/tensorflow/tensorflow/issues/30418#issuecomment-657101055), etc) are really saving GPU memory as intended. This is because I don't know what's the right tool to tell the difference in the results.

These are my questions: (all under the assumption that tensorflow 2.x is used, not 1.x versions)

1. How to properly profile the GPU memory usage of tensorflow? I tried pidajay's method, where the unmaintained/buggy(negative memory in the report, e.g. [issue](https://github.com/pythonprofilers/memory_profiler/issues/317#issue-896895819)) [pythonprofilers/memory_profiler](https://github.com/pythonprofilers/memory_profiler) is used. From [a reply by the author](https://github.com/pythonprofilers/memory_profiler/issues/200#issuecomment-398581049), it seems that the Repo. is for profiling CPU memory only (and pidajay also said he/she hadn't done the experiment for GPU/TPU [in some PR for tensorflow](https://github.com/tensorflow/tensorflow/issues/38766#issue-604279368) - see the 2. of the link).
2. So instead, I decided to use tensorboard to profile my GPU memory usage. I followed the idea of [the ipynb-tutorial by pidajay](https://github.com/pidajay/tf2_gradient_checkpointing/blob/master/tf_recompute_grad_tutorial.ipynb) that for each block/segment(containing many existing `tf.keras.layers`) to be recomputed, it should be wrapped within a `tf.keras.Sequential` for the current `@tf.recompute_grad` API to work). Then I tried many Repo.(auhors and links mentioned above) on GitHub for the API part. The following is a pair of two different reports from thensorboard that seems to mean the official API is working, but I cannot confirm it:

    - Fig 1.: The `@tf.recompute_grad` is not applied. The report has more rows than Fig 2.
    - Fig 2.: The `@tf.recompute_grad` is applied. But I don't know whether those `Conv2DBackpropFilter` would represent those call signatures and parameters saved during the forward pass for the recomputation on the backward pass.

<img width="1435" alt="image" src="https://user-images.githubusercontent.com/24765272/185365243-d9631f1d-0dbd-4914-8293-6470913d342a.png">

![image](https://user-images.githubusercontent.com/24765272/185365341-7516611d-6336-49fd-bfec-5ab51b3d6cdb.png)

### Standalone code to reproduce the issue

In every experiment, I divided VGG11/13/16/19 into `Segment`s with the following code with/without the `tf.recompute_grad` wrapper.

```python
import tensorflow as tf


class Segment(tf.keras.layers.Layer):
    def __init__(self, ll, **kwargs):
        super().__init__(**kwargs)
        self._layers = tf.keras.Sequential()
        for l in (ll):
            if l == 'M':
                self._layers.add(tf.keras.layers.MaxPool2D(pool_size=2, strides=2))
            else:
                self._layers.add(tf.keras.layers.Conv2D(l, kernel_size=3, padding='same'))
                self._layers.add(tf.keras.layers.BatchNormalization())
                self._layers.add(tf.keras.layers.ReLU())
        self._layers.add(tf.keras.layers.AveragePooling2D(pool_size=1, strides=1))

    def call(self, x):
        return tf.recompute_grad(self._layers)(x)
```

```python
import tensorflow as tf

from models.segment import Segment


config = {
    'vgg_new11': [64, 'M', 128, 'M', 256, 256, 'M', 512, 512, 'M', 512, 512, 'M'],
    'vgg_new13': [64, 64, 'M', 128, 128, 'M', 256, 256, 'M', 512, 512, 'M', 512, 512, 'M'],
    'vgg_new16': [64, 64, 'M', 128, 128, 'M', 256, 256, 256, 'M', 512, 512, 512, 'M', 512, 512, 512, 'M'],
    'vgg_new19': [64, 64, 'M', 128, 128, 'M', 256, 256, 256, 256, 'M', 512, 512, 512, 512, 'M', 512, 512, 512, 512, 'M'],
}



class VGG(tf.keras.Model):
    def __init__(self, vgg_name, num_classes):
        super().__init__()
        model = config[vgg_name]
        self.blocks = tf.keras.Sequential()
        _seg = []
        for i, l in enumerate(model):
            _seg.append(l)
            if (i+1) % 3 == 0:
                self.blocks.add(Segment(_seg))
                _seg = []
            if (i+1) == len(model) and len(_seg) > 0:
                self.blocks.add(Segment(_seg))

        self.flatten = tf.keras.layers.Flatten()
        self.fc = tf.keras.layers.Dense(num_classes, activation='softmax')

    def call(self, x, **kwargs):
        out = self.blocks(x)
        out = self.flatten(out)
        out = self.fc(out)
        return out
```

### Relevant log output

_No response_</details>

**Comment by nyngwang** [comment]
I would also suggest reopening issue #30418, because if you follow the thread you would know that the problem isn't resolved:

1. The thread got closed [after @ppham27 had shared the code from his research](https://github.com/tensorflow/tensorflow/issues/30418#issuecomment-914245598) and the author thumbed it. But from [the comment](https://github.com/tensorflow/tensorflow/issues/30418#issuecomment-1184185171) it seems that it doesn't work on tensorflow 2.9.1:

      > @ppham27 Hello! I've been trying to use your gradient checkpointing solution in tf 2.9.1 with jit compilation enabled, but keep running into an OOM that I do not face if I don't enable gradient checkpointing. I'm using a double stacked model: as in, I have a VAE with a separate encoder and decoder model. Do you have any easier-to-read docs for your gradient checkpointing py, and have you verified that it works on 2.9.1?

2. I have tried all the Repos in the thread, none of them work. Even if some of it might work, since I might monitor the GPU memory usage incorrectly, the documentation of the current `recompute_grad` API can be improved so people know how to correctly apply it so that the current `recompute_grad` can achieve  the same thing as what Pytorch [`torch.utils.checkpoint.checkpoint`](https://pytorch.org/docs/stable/checkpoint.html) does.

**Comment by nyngwang** [comment]
@gowthamkpr Could you provide some feedback...? If anything is not clear.

**Comment by nyngwang** [comment]
:(

**Comment by aniruthraj** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
